### PR TITLE
Swarm network rewrite syncer

### PR DESCRIPTION
**Pull request by zelig**
early PR for all to see what we are working on during this hackweek
closes https://github.com/ethersphere/go-ethereum/issues/181
 
see https://github.com/ethersphere/swarm/wiki/Light-mode-of-operation
and https://github.com/ethersphere/swarm/wiki/

### CODE REVIEW

File: swarm/network/stream/stream.go [hunk 1]
```diff
@@ -0,0 +1,441 @@
...
+	var total int64
+	var err error
+	for (total == 0 || n > 0) && err == nil {
+		n, err = r.ReadAt(buf, total)
+		total += int64(n)
+	}
+	if err != nil && err != io.EOF {
+		return total, err
+	}
+	return total, nil
```
Comment (lmars) [inline swarm/network/stream/stream.go:428]: If the idea is to read all the data and return the number of bytes read, perhaps use `io.Copy(ioutil.Discard, r)`?

File: swarm/network/stream/messages.go [hunk 2]
```diff
@@ -0,0 +1,277 @@
...
+	// go p.SendOfferedHashes(s, req.From, req.To)
+	l := len(hashes) / HashSize
+	want, err := bv.NewFromBytes(req.Want, l)
+	if err != nil {
+		return fmt.Errorf("error initiaising bitvector of length %v: %v", l, err)
+	}
+	for i := 0; i < l; i++ {
+		if want.Get(i) {
+			hash := hashes[i*HashSize : (i+1)*HashSize]
+			data, err := s.GetData(hash)
```
Comment (lmars) [inline swarm/network/stream/messages.go:224]: This will block the peer run loop (so potentially preventing the handling of higher priority requests), should it be placed in a goroutine?

File: swarm/network/stream/stream.go [hunk 3]
```diff
@@ -0,0 +1,441 @@
...
+		delivery:    delivery,
+	}
+	streamer.api = NewAPI(streamer, streamer.store)
+	delivery.getPeer = streamer.getPeer
+	streamer.RegisterServerFunc(swarmChunkServerStreamName, func(_ *Peer, t []byte) (Server, error) {
+		return NewSwarmChunkServer(delivery.db), nil
+	})
+	streamer.RegisterClientFunc(swarmChunkServerStreamName, func(p *Peer, t []byte) (Client, error) {
+		return NewSwarmSyncerClient(p, delivery.db, nil)
+	})
```
Comment (lmars) [inline swarm/network/stream/stream.go:79]: I find it confusing that `NewRegistry` registers a syncer client and a chunk server, but not a syncer server? Is there a reason for this?

It's a little hard to follow how `Registry` / `Delivery` and the various `Client` / `Server` implementations all fit together, it would be good to have some high level documentation.

**Review by lmars — COMMENTED** [review]
I've tried to review the best I can given the size (I skipped reviewing any tests). Some minor comments and clarifications, but largely looks solid :+1:

**Comment by nonsense** [comment]
Having spent my recent efforts on `swarm-network-rewrite-syncer`, I will look into resolving the conflicts with `swarm-network-rewrite`, and bringing this PR to a merge-able state.
